sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1118,6 +1119,99 @@ ProgramManager::KernelArgMask ProgramManager::getEliminatedKernelArgMask(
   return {};
 }
 
+static bundle_state getBinImageState(RTDeviceBinaryImage *BinImage) {
+  auto IsAOTBinary = [](const char *Format) {
+    return (
+        (strcmp(Format, __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_X86_64) == 0) ||
+        (strcmp(Format, __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_GEN) == 0) ||
+        (strcmp(Format, __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_FPGA) == 0));
+  };
+
+  // There are only two initial states so far - SPIRV which needs to be compiled
+  // and linked and fully compiled(AOTed) binary
+
+  const bool IsAOT = IsAOTBinary(BinImage->getRawData().DeviceTargetSpec);
+
+  return IsAOT ? sycl::bundle_state::executable : sycl::bundle_state::input;
+}
+
+static bool compatibleWithDevice(RTDeviceBinaryImage *BinImage,
+                                 const device &Dev) {
+  const std::shared_ptr<detail::device_impl> &DeviceImpl =
+      detail::getSyclObjImpl(Dev);
+  auto &Plugin = DeviceImpl->getPlugin();
+
+  const RT::PiDevice &PIDeviceHandle = DeviceImpl->getHandleRef();
+
+  // Call piextDeviceSelectBinary with only one image to check if an image is
+  // compatible with implementation. The function returns invalid index if no
+  // device images are compatible.
+  pi_uint32 SuitableImageID = 42;
+  pi_device_binary DevBin =
+      const_cast<pi_device_binary>(&BinImage->getRawData());
+  Plugin.call<PiApiKind::piextDeviceSelectBinary>(PIDeviceHandle, &DevBin,
+                                                  /*Image size = */ (cl_uint)1,
+                                                  &SuitableImageID);
+  const bool Compatible = (0 == SuitableImageID);
+
+  return Compatible;
+}
+
+std::vector<device_image_plain>
+ProgramManager::getSYCLDeviceImages(const context &Ctx,
+                                    const std::vector<device> &Devs,
+                                    bundle_state TargetState) {
+
+  // Collect raw device images
+  std::vector<RTDeviceBinaryImage *> BinImages;
+  {
+    std::lock_guard<std::mutex> Guard(Sync::getGlobalLock());
+    for (auto &ImagesSets : m_DeviceImages) {
+      auto &ImagesUPtrs = *ImagesSets.second.get();
+      for (auto &ImageUPtr : ImagesUPtrs)
+        BinImages.push_back(ImageUPtr.get());
+    }
+  }
+
+  // Create SYCL device image from those that have compatible state and at least
+  // one device
+  std::vector<device_image_plain> SYCLDeviceImages;
+  for (RTDeviceBinaryImage *BinImage : BinImages) {
+    const bundle_state ImgState = getBinImageState(BinImage);
+    // Ignore images with incompatible state - image is in "executable" state
+    // while kernel_bundle needs to be in "input" state
+    if (ImgState > TargetState)
```

Comment:
Please explain in the comment what "compatible" really means. If I guess correctly, if a binary image in state A can be brought to state B, then A is "compatible" with B. (maybe "compatible" is not a good term actually, as the relation is not symmetric, while the term kind of implies symmetry)

Reply:
Improved the comment.